### PR TITLE
date time range assertions - asserting difference is not negative

### PR DESCRIPTION
**Pull request by jonathann92**
## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution affects [the documentation](https://github.com/fluentassertions/fluentassertions/tree/master/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com).

Fixing issue where assertion on `DateTimeRange` will fail on `BeLessThan` or `BeWithin` when the subject and target has a negative delta.

Reported in #1165

### CODE REVIEW

**Comment by jnyrup** [comment]
Thanks for looking in this issue 👍 

I like the idea about replacing `differs` with the explicit `before`/`after`.
It's more human-friendly to read `before by 10s`, than `differs by -10s`.

Could this be simplified by moving the `actual >= TimeSpan.Zero` into the dictionary of `predicates` 
E.g.
```c#
ts1 >= TimeSpan.Zero && ts1 > ts2
```

Then if `IsMatched` fails, whether `actual` is positive or negative can control whether the failure message should be `but it is after` or `but it is before`.

File: Src/FluentAssertions/Primitives/DateTimeOffsetRangeAssertions.cs [hunk 1]
```diff
@@ -112,15 +119,22 @@ public AndConstraint<TAssertions> After(DateTimeOffset target, string because =
...
-                            "Expected {context:the date and time} to be " + predicate.DisplayText +
-                            " {0} after {1}{reason}, but {2} differs {3}.",
-                            timeSpan, target, subject, actual);
-                }
+                Execute.Assertion
+                    .ForCondition(actual >= TimeSpan.Zero)
+                    .BecauseOf(because, becauseArgs)
+                    .FailWith(
+                        "Expected {context:the date and time} {0} to be " + predicate.DisplayText +
+                        " {1} after {2}{reason}, but it is before by {3}.",
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/DateTimeOffsetRangeAssertions.cs:127]: I think it should be something like _but it is behind by_.

File: Src/FluentAssertions/Primitives/DateTimeRangeAssertions.cs [hunk 2]
```diff
@@ -75,15 +75,22 @@ public class DateTimeRangeAssertions<TAssertions>
...
-                            "Expected date and/or time {0} to be " + predicate.DisplayText +
-                            " {1} before {2}{reason}, but it differs {3}.",
-                            subject, timeSpan, target, actual);
-                }
+                Execute.Assertion
+                    .ForCondition(actual >= TimeSpan.Zero)
+                    .BecauseOf(because, becauseArgs)
+                    .FailWith(
+                        "Expected {context:the date and time} {0} to be " + predicate.DisplayText +
+                        " {1} before {2}{reason}, but it is after by {3}.",
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/DateTimeRangeAssertions.cs:83]: Instead of _after by_, use _ahead by_

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs [hunk 3]
```diff
@@ -1902,6 +1902,144 @@ public void When_time_is_less_than_30s_after_another_time_it_should_not_throw()
             subject.Should().BeLessThan(TimeSpan.FromSeconds(30)).After(target);
         }
 
+        [Fact]
+        public void When_asserting_subject_be_morethan_10_seconds_after_target_but_subject_is_before_target_should_throw()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:1906]: (Nitpicking) Missing _  between _more_ and _than_

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs [hunk 4]
```diff
@@ -1902,6 +1902,144 @@ public void When_time_is_less_than_30s_after_another_time_it_should_not_throw()
...
+        {
+            // Arrange
+            var target = 1.January(0001).At(0, 0, 30);
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:1909]: Instead of _target_ we use the term _expectation_ everywhere.

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs (hunk 3 above)
Comment (jonathann92) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:1906]: I was debating if i should have a `be_more_than` or `bemorethan` because the method name is `BeMoreThan` but somehow I went in the middle lol. I'll put an underscore to between the words.

**Comment by jonathann92** [comment]
@dennisdoomen 

> I think it should be something like _but it is behind by_.

> Instead of _after by_, use _ahead by_

My reasoning for using `before` and `after` is that its the same wording as the method name. 

Also ahead can mean "before" in some contexts other than time and this can lead to confusion. For example, there are 2 people in a line: 

- The 1st person is _before_ the 2nd person
- The 1st person is _ahead_ of the 2nd person

**Comment by dennisdoomen** [comment]
> _but it is before by 2 ms_

I'm not a native speaker, but somehow it feels wrong. I'm wondering what is the _correct_ English wording. Giving you're from CA, I guess you should know better than me.

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs (hunk 3 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:1906]: > Don’t refer to technical names such as the names of methods, variables and UI elements in the name of a test. In my opinion, the name of a test should explain the functional scenario, whereas the cause and effect should clearly illustrate the technicalities of that scenario. If some class or member is renamed, it shouldn’t affect the functional name of the test.

From https://www.continuousimprover.com/2015/11/12-tips-to-write-unit-tests-that-dont.html

In other words, I never use the literal name of a specific method in a test name.

**Comment by jnyrup** [comment]
The no native speaker disclaimer applies to me as well.

"but it is before by 2 ms" sounds like a mix of
"but it is behind by 2 ms" and "but it is 2 ms before".

I think behind/ahead is used with time if e.g. your watch is out of sync, then 
"it is behind by 2s".

**Comment by jonathann92** [comment]
Ok it makes sense when we're talking about times. I changed the messaging to ahead/behind instead of after/before.

**Comment by dennisdoomen** [comment]
Are you willing to also update the [release notes](https://github.com/fluentassertions/fluentassertions/blob/develop/docs/_pages/releases.md#600)?

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs [hunk 5]
```diff
@@ -1902,6 +1902,144 @@ public void When_time_is_less_than_30s_after_another_time_it_should_not_throw()
...
+            var subject = 1.January(0001).At(0, 0, 15);
+
+            // Act / Assert
+            Action action = () => subject.Should().BeMoreThan(10.Seconds()).After(target);
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:1913]: > We do use target elsewhere in DateTimeOffsetAssertionSpecs.cs

@jnyrup legacy code ;-)

File: Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs (hunk 4 above)
Comment (jonathann92) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeAssertionSpecs.cs:1909]: @jnyrup I can change `target` to `expectation` in the files I touched. Would you guys also like to me to change it in `DateTimeOffsetRangeAssertions.cs` and `DateTimeOffsetRangeAssertions.cs`?

**Comment by jonathann92** [comment]
> Are you willing to also update the [release notes](https://github.com/fluentassertions/fluentassertions/blob/develop/docs/_pages/releases.md#600)?

Does this wording sound right?
```Asserting DateTime(Offset) `Before` will fail when subject is after and vice versa```

**Comment by jnyrup** [comment]
@jonathann92 I would write it in past tense, since it's now fixed.
@dennisdoomen Any comments on how to formulate this in the release notes?

**Comment by dennisdoomen** [comment]
> Any comments on how to formulate this in the release notes?

Taking from the original bug report:

    Ensures that date time assertions like "a is less than an hour after b" don't succeed when `a - b == -30.Minutes()`

**Comment by jnyrup** [comment]
@jonathann92 again, thanks for taking the time to look this issue and contributing a fix ❤️

**Comment by dennisdoomen** [comment]
Kudos for @jonathann92 indeed. With all these contributions, 6.0 is going to be awesome.
